crate/operator/main.py

```diff
@@ -386,188 +379,56 @@ async def cluster_create(
-    do_upgrade = False
-    requires_restart = False
-    scale_master_diff_item: Optional[kopf.DiffItem] = None
-    scale_data_diff_items: Optional[List[kopf.DiffItem]] = None
+    context = Context.deserialize(status.get("updateContext"))
+    new_cycle = False
+    if context.state_machine.done:
+        new_cycle = True
 
-    for operation, field_path, old_value, new_value in diff:
+    do_upgrade = False
+    do_restart = False
+    do_scale = False
+    for _, field_path, *_ in diff:
         if field_path in {
             ("spec", "cluster", "imageRegistry"),
             ("spec", "cluster", "version"),
         }:
             do_upgrade = True
+            do_restart = True
         elif field_path == ("spec", "nodes", "master", "replicas"):
-            do_scale_master = True
-            scale_master_diff_item = kopf.DiffItem(
-                operation, field_path, old_value, new_value
-            )
+            do_scale = True
         elif field_path == ("spec", "nodes", "data"):
-            # TODO: check for data node order, added or removed types, ...
-            if len(old_value) != len(new_value):
-                raise kopf.PermanentError(
-                    "Cannot handle changes to the number of node specs."
-                )
-            scale_data_diff_items = []
-            for node_spec_idx in range(len(old_value)):
-                old_spec = old_value[node_spec_idx]
-                new_spec = new_value[node_spec_idx]
-                inner_diff = calc_diff(old_spec, new_spec)
-                for (
-                    inner_operation,
-                    inner_field_path,
-                    inner_old_value,
-                    inner_new_value,
-                ) in inner_diff:
-                    if inner_field_path == ("replicas",):
-                        do_scale_data = True
-                        scale_data_diff_items.append(
-                            kopf.DiffItem(
-                                inner_operation,
-                                (str(node_spec_idx),) + inner_field_path,
-                                inner_old_value,
-                                inner_new_value,
-                            )
-                        )
-                    else:
-                        logger.info(
-                            "Ignoring operation %s on field %s",
-                            operation,
-                            field_path + (str(node_spec_idx),) + inner_field_path,
-                        )
-        else:
-            logger.info("Ignoring operation %s on field %s", operation, field_path)
+            do_scale = True
 
-    async with ApiClient() as api_client:
-        apps = AppsV1Api(api_client)
-        coapi = CustomObjectsApi(api_client)
-        core = CoreV1Api(api_client)
+    if do_upgrade:
+        if new_cycle:
+            context.state_machine.add(State.UPGRADE)
+        kopf.register(fn=UpgradeSubHandler(namespace, name, context)(), id="upgrade")
```

Comment:
I find it a bit unconventional that you need to call the instantiated subhandler instance to get the subhandler function. I understand that it comes from `__call__` being a regular method, but the handler is an async method. The option to pass `UpgradeSubHandler(namespace, name, context)._subhandler` isn't pretty either.

Reply:
As noted above, I originally planned on only passing the sub-handler instance and have a `async def __call_(self, **kwargs)` which does what `async def _subhandler(self, **kwargs)` does. However, that causes some errors inside serialization in Kopf which I didn't investigate further.